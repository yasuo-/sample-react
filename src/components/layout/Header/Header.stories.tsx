import React from 'react';

import { ComponentStory, ComponentMeta } from '@storybook/react';

import { Header } from './Header';

export default {
  title: 'Components/Layout/Header',
  component: Header,
  parameters: {
    layout: 'fullscreen',
  },
} as ComponentMeta<typeof Header>;

const Template: ComponentStory<typeof Header> = (args) => <Header {...args} />;

export const Default = Template.bind({});

Default.args = {
  siteTitle: 'logo',
  src: 'https://via.placeholder.com/150',
  logoWidth: 150,
  logoHeight: 50,
};
