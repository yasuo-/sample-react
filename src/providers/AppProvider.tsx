import React from 'react';

import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { ErrorBoundary } from 'react-error-boundary';
import { HelmetProvider, HelmetData } from 'react-helmet-async';
import { I18nextProvider } from 'react-i18next';

import { ErrorFallback } from '@/components/error';
import i18n from '@/lib/locale/i18n';

type AppProviderProps = {
  children: React.ReactNode;
};

// Create a client
const queryClient = new QueryClient();
const helmetContext = new HelmetData({});

/**
 * AppProvider
 * @param children
 * @returns {JSX.Element}
 */
export const AppProvider: React.FC<AppProviderProps> = ({ children }) => {
  return (
    <I18nextProvider i18n={i18n}>
      <QueryClientProvider client={queryClient}>
        <HelmetProvider context={helmetContext}>
          <ErrorBoundary FallbackComponent={ErrorFallback}>{children}</ErrorBoundary>
        </HelmetProvider>
      </QueryClientProvider>
    </I18nextProvider>
  );
};
