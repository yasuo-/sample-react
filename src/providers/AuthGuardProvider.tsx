import React, { createContext, useContext } from 'react';

export type AuthGuardContextType = {
  accessToken: string | null;
  signIn: (accessToken: string) => void;
  signOut: () => void;
};

type AuthGuardProviderProps = {
  children: React.ReactNode;
};

/**
 * AuthGuardContext
 */
const AuthGuardContext = createContext<AuthGuardContextType>({} as AuthGuardContextType);

/**
 * useAuthGuardContext
 */
export const useAuthGuardContext = (): AuthGuardContextType =>
  useContext<AuthGuardContextType>(AuthGuardContext);

/**
 * AuthGuardProvider
 * @param AuthGuardProviderProps {children: React.ReactNode}
 * @constructor
 */
export const AuthGuardProvider: React.FC<AuthGuardProviderProps> = ({ children }) => {
  const [accessToken, setAccessToken] = React.useState<string | null>(null);
  /**
   * signIn
   * @param accessToken {string} access token
   */
  const signIn = (accessToken: string) => {
    setAccessToken(accessToken);
  };
  /**
   * signOut
   */
  const signOut = () => {
    setAccessToken(null);
  };

  return (
    <AuthGuardContext.Provider value={{ accessToken, signIn, signOut }}>
      {children}
    </AuthGuardContext.Provider>
  );
};
