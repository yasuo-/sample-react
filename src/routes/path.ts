export const PATH = {
  HOME: '/',
  // Auth
  SIGNIN: '/signin',
  SIGNUP: '/signup',
  FORGOT_PASSWORD: '/forgot-password',
  RESET_PASSWORD: '/reset-password',
  // Dashboard
  DASHBOARD: '/dashboard',
  ACCOUNT: '/account',

  // Common
  TERMS: '/terms',
  PRIVACY: '/privacy-policy',
  CONTACT: '/contact',
  // Error
  ERROR: '/error',
};
