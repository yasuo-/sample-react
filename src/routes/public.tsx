import { Navigate } from 'react-router-dom';

import { PublicLayout } from '@/components/layout';

export const publicRoutes = [
  {
    path: '/',
    element: <PublicLayout />,
    children: [
      { path: '/', element: <></> },
      { path: '/error', element: <></> },
      { path: '*', element: <Navigate to="." /> },
    ],
  },
];
