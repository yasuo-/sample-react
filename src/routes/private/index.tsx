import React from 'react';

import { Navigate } from 'react-router-dom';

import { PATH } from '../path';

import type { RouteObject } from 'react-router-dom';

import { ErrorNotFound } from '@/components/error';
import { AuthGuardLayout } from '@/components/layout';
import { fetchAccount } from '@/pages/Account/Account';
import { queryClient } from '@/providers';

const AccountLayout = React.lazy(() => import('@/pages/Account/Account'));
const AccountPage = React.lazy(() => import('@/pages/Account/AccountIndex'));
const AccountEditPage = React.lazy(() => import('@/pages/Account/AccountEdit'));
/**
 * interface RouteObject {
 *   path?: string;
 *   index?: boolean;
 *   children?: React.ReactNode;
 *   caseSensitive?: boolean;
 *   id?: string;
 *   loader?: LoaderFunction;
 *   action?: ActionFunction;
 *   element?: React.ReactNode | null;
 *   errorElement?: React.ReactNode | null;
 *   handle?: RouteObject["handle"];
 *   shouldRevalidate?: ShouldRevalidateFunction;
 * }
 */

export const privateRoutes: RouteObject[] = [
  {
    path: '/',
    element: <AuthGuardLayout />,
    errorElement: <ErrorNotFound />,
    loader: fetchAccount(queryClient),
    children: [
      { index: true, element: <>mypage</> },
      {
        path: PATH.ACCOUNT,
        element: <AccountLayout />,
        children: [
          { index: true, element: <AccountPage /> },
          { path: PATH.ACCOUNT_EDIT, element: <AccountEditPage /> },
        ],
      },
      // Error
      { path: PATH.ERROR, element: <></> },
      { path: '*', element: <Navigate to="." /> },
    ],
  },
];
