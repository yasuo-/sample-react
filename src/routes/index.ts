import { useRoutes, createBrowserRouter } from 'react-router-dom';

import { privateRoutes } from './private';
import { publicRoutes } from './public';

/**
 * AppRoutes
 * react-router-dom v6 <BrowserRouter>
 */
export const AppRoutes = () => {
  // auth logic
  const auth = false;
  const routes = auth ? privateRoutes : publicRoutes;
  return useRoutes([...routes]);
};

/**
 * router
 * react-router-dom v6.4~ <RouterProvider>
 */
export const router = createBrowserRouter([...publicRoutes, ...privateRoutes]);
