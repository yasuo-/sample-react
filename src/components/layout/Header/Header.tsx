import React from 'react';

import { Link } from 'react-router-dom';

import { PATH } from '@/routes/path';

type HeaderProps = {
  siteTitle?: string;
  src?: string;
  logoWidth?: number;
  logoHeight?: number;
};

/**
 * Header
 * @param alt
 * @param src
 * @param logoWidth
 * @param logoHeight
 * @returns Header
 */
export const Header: React.FC<HeaderProps> = ({
  siteTitle = 'logo',
  src = 'https://via.placeholder.com/150',
  logoWidth = 150,
  logoHeight = 50,
}) => {
  return (
    <header className="w-full bg-fuego-500">
      <div className="flex justify-center items-center h-10">
        <h1>
          <Link to={PATH.HOME} className="hover:opacity-75">
            <img src={src} width={logoWidth} height={logoHeight} alt={siteTitle} />
          </Link>
        </h1>
      </div>
    </header>
  );
};

Header.displayName = 'Header';
