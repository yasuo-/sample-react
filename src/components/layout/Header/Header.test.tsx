import React from 'react';

import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { MemoryRouter } from 'react-router-dom';

import { Header } from './Header';

import { AppProvider } from '@/providers';
import { PATH } from '@/routes/path';

/**
 * renderWithRouter
 * helper function to render component with router
 * @param component
 */
const renderWithRouter = (component: React.ReactNode) => {
  return {
    ...render(
      <AppProvider>
        <MemoryRouter>{component}</MemoryRouter>
      </AppProvider>
    ),
  };
};

describe('Header Components', () => {
  it('displayName is Header', () => {
    expect(Header.displayName).toBe('Header');
  });

  it('should take a snapshot', () => {
    const { asFragment } = renderWithRouter(<Header siteTitle={'test title'} />);
    expect(asFragment).toMatchSnapshot();
  });

  it('renders test title', () => {
    renderWithRouter(<Header siteTitle={'test title'} />);
    const screenElement = screen.getByAltText(/test title/i);
    expect(screenElement).toBeInTheDocument();

    userEvent.click(screenElement);
    expect(window.location.pathname).toEqual(PATH.HOME);
  });
});
