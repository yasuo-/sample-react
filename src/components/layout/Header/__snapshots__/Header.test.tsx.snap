// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Components should take a snapshot 1`] = `[Function]`;
